### ISSUE TITLE
Improve memory allocation for LLM-like models

### ISSUE THREAD

**Issue by silvasean** [iree-org/iree#13545]
### What happened?

Same repro as https://github.com/openxla/iree/issues/13543, but patch https://github.com/openxla/iree/blob/03110def3fab268812ee33b183786fd991514a68/runtime/src/iree/modules/hal/module.c#L1121 to increase the count from 32 to 128 to work around #13543.

Error:

```
iree/runtime/src/iree/hal/drivers/cuda/cuda_device.c:538: INTERNAL; CUDA driver error 'CUDA_ERROR_ILLEGAL_ADDRESS' (700):
 an illegal memory access was encountered; cuStreamSynchronize; while invoking native function hal.device.queue.execute; 
while calling import;
```

(full error log [here](https://gist.github.com/silvasean/07b8a41555412804e4d16143b124d415))

### Steps to reproduce your issue

See above

### What component(s) does this issue relate to?

Runtime

### Version information

iree.git @ https://github.com/openxla/iree/commit/4d6c2b8699383598817352f19b5888b38be06ccb (I have a couple trivial local modifications, but should be unrelated to this bug)

**Comment by silvasean** [comment]
If I comment out the cuStreamSynchronize then the error gets attributed to:

```
iree/runtime/src/iree/hal/drivers/cuda/stream_command_buffer.c:507: INTERNAL; CUDA driver error 'CUDA_ERROR_ILLEGAL_ADDRESS' (700): an illegal memory access was encountered; cuLaunchKernel; while invoking native function hal.device.queue.execute; while calling import;
```

**Comment by silvasean** [comment]
I suspect that this is due to a miscompile. Can somebody confirm that that is what this likely means? I will bisect it tomorrow.

**Comment by silvasean** [comment]
It looks like it is likely due to running out of memory. It would be nice if we could diagnose this better. I will try a smaller variant of this model. However, it is important to note that this model runs successfully on XLA:GPU with the available memory (A100-40GB), so perhaps we can use this bug to guide improvements to our memory allocation strategy. I renamed the bug to reflect that.

**Comment by silvasean** [comment]
I tried with `--device_allocator=caching ` and got the same error.

**Comment by silvasean** [comment]
Also if this is IREE's OOM failure mode, we should try to improve this, as it is really non-obvious. XLA:GPU gives a detailed diagnosis listing the buffers involved in the OOM situation and their sizes and names.

**Comment by kushanam** [comment]
@silvasean I tried to repro the issue with the latest iree commit and I get 
`LLVM ERROR: operation destroyed but still has uses` while creating the vmfb file. Any ideas what might have I done wrong?

**Comment by stellaraccident** [comment]
That looks like the thing that Sean was working around by disabling the transform dialect for CUDA. I can't find the issue right now, but it is waiting until next week for the owner of the component to return from vacation.

**Comment by stellaraccident** [comment]
This one, I think: https://github.com/openxla/iree/issues/13419

**Comment by kushanam** [comment]
Thank you @stellaraccident!

**Comment by kushanam** [comment]
@stellaraccident seems like I already tried the workaround which is also in the linked issue. Still getting the same error.

**Comment by silvasean** [comment]
That is a separate issue tracked here: https://github.com/openxla/iree/issues/13459
I don't know why I am not hitting it in my particular setup.

**Comment by frgossen** [comment]
As Kushan had asked me, I am seeing the same error on `iree-compile`

**Comment by silvasean** [comment]
It looks like the fix for #13459 landed today: https://github.com/openxla/iree/pull/13625
Hopefully it should work now at head.

**Comment by benvanik** [comment]
Can you post a trace of the execution? It'll be easy to see memory consumption over time and then match it up to the model to figure out what patterns could be improved/worked around. Also, please post a copy of the StableHLO IR so that we can work through the issues and relate them to #13637.

(nothing here is CUDA specific - you're just using CUDA - renaming this because it's about the allocation patterns in the input model and not the backend)

**Comment by benvanik** [comment]
Fiddling, some initial suggestions:
`--iree-stream-partitioning-favor=min-peak-memory` will help a bit on these models. Getting state properly modeled in JAX is by far the largest improvement that can be done - these models are cycling through 10's of GB of state in ways we cannot really optimize well today. In-place outputs may help (I've got to experiment some more) but won't solve it all. Removing the loops if you don't need them (I think at some point I heard you always just loop once) will dramatically improve things.

**Comment by silvasean** [comment]
To reproduce locally, using [ir19.flagfile](https://gist.githubusercontent.com/silvasean/3cf5d0078a2351e601c7f46e2b173cd6/raw/ca9103a884b81d9c24aed4c4663780e05daa9500/ir19.flagfile) and [ir19.no_sharding.mlir](https://gist.githubusercontent.com/silvasean/3cf5d0078a2351e601c7f46e2b173cd6/raw/ca9103a884b81d9c24aed4c4663780e05daa9500/ir19.no_sharding.mlir), run the following commands:
```
iree-compile --iree-hal-target-backends=cuda --iree-input-type=mhlo --iree-hal-cuda-llvm-target-arch=sm_80 ir19.no_sharding.mlir --iree-codegen-llvmgpu-use-transform-dialect= --iree-codegen-llvmgpu-enable-transform-dialect-jit=false -o ir19.vmfb
iree-benchmark-module --function=main --module=ir19.vmfb --device=cuda --flagfile=ir19.flagfile --benchmark_repetitions=5
```


The baseline I have for this (from jax running on XLA:GPU) uses 10GiB peak memory. We will need to identify steps to reach that level. Currently IREE appears to be using at least 25GiB.

At the moment I don't have enough time allocated to this issue to do a deeper memory profile myself. We may want to meta-bug this with https://github.com/openxla/iree/issues/13566 as "efficient JAX end-to-end memory allocation story" which needs an owner.

**Comment by stellaraccident** [comment]
@allieculp Can we get this and #13566 on top issues tomorrow?

**Comment by allieculp** [comment]
@julianwa @jpienaar @mattwalsh to find an owner for this project.

**Comment by allieculp** [comment]
Updates from offline conversations: We have a potential owner candidate, work to start ~July. Some spot fixes until then.

**Comment by stellaraccident** [comment]
I would feel a lot better about a delayed start on overall work here if we at least had some scope/knowns called out. Can be rough, but not super comfortable with unknown-unknowns.

**Comment by benvanik** [comment]
Currently working on improving the iree-stream-emplace-allocations pass, which should help a bit when completed (~27gb to ~8gb, on whatever LLM slice I'm looking at). Only have 27->26gb now but have visibility on 26->8. I'll have a good weekend if I can get an order of magnitude improvement :)

I think the next step after that is to induce more in-place dispatches during dispatch region formation. Some of these individual tensors I'm seeing are 8+gb, so even one copy is fairly catastrophic. We seem to be doing a somewhat decent job (and my fixes above are mostly taking advantage of the dispatch regions that have in-place behavior we don't currently support in that pass) and we'll have a better grasp on what other opportunities there are after the fixes are in. @MaheshRavishankar's work on improving small scalar fusions will also reduce a lot of the noise that is making it harder to track.

**Comment by MaheshRavishankar** [comment]
Sounds good. I was on track for sending something out today, but had to take the day off. I'll send it out next week. I'll sync up with Ben about what he is seeing to make progress on these

**Comment by silvasean** [comment]
Awesome Ben!

> I think the next step after that is to induce more in-place dispatches during dispatch region formation. Some of these individual tensors I'm seeing are 8+gb, so even one copy is fairly catastrophic. 

@benvanik where are you seeing 8GB tensors? (what is the shape so I can grep the program for them -- or did you mean MB?) If my regex is right, the largest tensors I see in the model I think should be the attention logits which are `batch=1 * num_heads=32 * sequence_len=2048 * sequence_len=2048 x f32 = ~500MB` -- note that these are also the intermediate tensors that the FlashAttention fusion will eliminate.

**Comment by benvanik** [comment]
In the thing I'm looking at in #13648 there's many large tensors returned as outputs:
```
402653184
402653184
402653184
419430400
419430400
419430400
1207959552
1207959552
1207959552
1610612736
1610612736
1610612736
1610612736
1610612736
1610612736
```
Totals 15810161504 (~16GB) of exported tensors, and some of those overlap with the large inputs as well. The remaining 10GB allocated in the program comes from transients and extra allocations induced by the loop (which I think you had mentioned could be removed, and until there's time to eliminate those definitely should be to save many GB!).

**Comment by silvasean** [comment]
Ah yes, in a GPT-3 scale model indeed that is the size. I was worried if you saw such large individual tensors in the repro in this bug :)

**Comment by benvanik** [comment]
Due to issues with stablehlo.gather iree-compile can't currently accept any of the repros that have been shared for these. Seeing if anyone knows what's up in #13729.

**Comment by silvasean** [comment]
Just checking -- Ben, do you need me to regenerate the test case or something to unblock you here?

**Comment by stellaraccident** [comment]
> Just checking -- Ben, do you need me to regenerate the test case or something to unblock you here?

That would probably be the shortest way to get there. I'm also interested in stablehlo dev flows being more... stable. But I suspect any fixes there will be longer than just redoing these things.

**Comment by silvasean** [comment]
It sounded from the other bug that `--iree-input-type=mhlo` fixed the issue Ben was running into (vs `=stablehlo`). That's why I'm asking.

**Comment by benvanik** [comment]
yeah, I'll give this another go with the mhlo pipeline - I'd been assuming we were on stablehlo given this is all stablehlo IR but there's apparent differences.

**Comment by MaheshRavishankar** [comment]
I tried with `--iree-input-type=mhlo` and that seemed to work. I did use `--iree-input-type=stablehlo` and didnt work, and assumed that I was wrong in using that flag.

**Comment by stellaraccident** [comment]
Alright - I withdraw my attempt at directing traffic. Back to work :)

**Comment by MaheshRavishankar** [comment]
Some updates from me.... I think the first order of business here is to handle the operations that are zero-ranked.... 

- The more direct solution is to address #6948 . I have #13711  to implement something that handles it in a more direct way..
- But while looking at the IR for the input model here, and while chatting with @silvasean , he raised a good question of why our current elementwise operation fusion does not already collapse all these tensor operations into a single op. Digging a bit, this is because of the ops having multiple uses, and today those arent fused (being conservative). It is possible to do though, and is implemented under the flag `--iree-flow-fuse-multi-use` . It wasnt enabled by default cause it was too aggressive for the backends to handle, but I think a less aggressive version can be handled by default and should help a lot in this case (and other cases). So I am trying to enable that first in https://github.com/openxla/iree/pull/13747 
- I also notice there are a couple of ops with multiple reduction dimensions that are messing up some of the foldings. This was hit in a separate issue and https://github.com/openxla/iree/pull/13730 fixes that.

So in order of work for me here

- [x] Land #13730 (almost done)
- [ ] Investigate how much #13747 helps. If that gets us most of the way thats ideal cause it would mean any special handling for rank-zero operations
- [ ] In generality, we still need #13711 . Will come to this based on what happens with the other two.

**Comment by benvanik** [comment]
Have Sean's repro from #13543 working now and am going to work on the in-place update improvements. The major work after that I see is all CFG related and if we can get a copy of the model with the loops flattened (IIRC @silvasean you said the loop count was always constant 1 and it could be removed?) that'd be useful. Better allocations around CFG are a long-term workstream.

As a side-note, `--iree-opt-strip-assertions` would be a useful flag for JAX/PJRT models as each one currently does hundreds of shape/encoding/etc assertions.

**Comment by stellaraccident** [comment]
Thanks, Ben.

**Comment by silvasean** [comment]
Awesome.

In general for LLM training that loop will be >1 (say like 4), so it would be good for us to be robust to that. (though somehow I feel like it should canonicalize away at the arith level at least -- maybe core MLIR doesn't have some basic folds missing? -- regardless, we need to handle the non-trivial case).

**Comment by benvanik** [comment]
In the fullness of time for sure - for now, you're going to want to make sure that 4 comes through as a constant we can see if you don't want to OOM :)
(it may, too many models flying around, some with control flow and some without - will be good once we have the base changes here in to rally around a known-representative model with the right tweaks)

**Comment by silvasean** [comment]
Yes, the 4 will always be hardcoded (it is a model hyperparameter). Note that this sort of "microbatching" is also present in other workloads, like AlphaFold. The pattern here is that a pure batch dimension gets while-loop+slice'd over. This reduces the intermediate size of activations and in the distributed case allows more fine-grained producer-consumer chunks being passed to get some pipelineing in producer-consumer parallelism (see here: https://pytorch.org/docs/stable/pipeline.html).

**Comment by benvanik** [comment]
Goal is to support stuff like that - I'd love to have passes that did that too! For now we can unroll even if after flow and before we allocate in stream to still get the benefits. Major issue is that due to legacy MHLO reasons we bake out to CFG really early and it gets much harder to do the placements vs. when we have structured control flow. It'd be absolutely killer to have a CFG -> SCF pass in MLIR (ala https://dl.acm.org/doi/pdf/10.1145/3547621) so we could be robust to inputs in different modes. If you're looking for a compiler pass to write... ;)

**Comment by jpienaar** [comment]
> for LLM training

Do we have a similar profile for the inference workload and is it exhibiting the same characteristics as the training one?
 
> Major issue is that due to legacy MHLO reasons we bake out to CFG really early

Do we have these tracked?

**Comment by benvanik** [comment]
Damn, have some good improvements but none impact the memory consumption of Sean's particular model (with `--iree-stream-partitioning-favor=min-peak-memory` set) and instead only save a few GB of copies. The issues are still the control flow and most critically all of the out-of-place dispatches that lead to higher peak memory consumption - the transient usage there hides all of the improvements I could make at the stream level (well, not all - one transient buffer went from 1426063488 bytes to 1426063424 bytes :).

After I land these improvements (they clean up the IR quite a bit at the stream level) I can start digging into control flow to scope that out (probably "big"). TopLevelSCFToCFG is the pass that does SCF->CFG at the MHLO level today. #13542 mentions some things we can do but we still need to scope out the work as there's a lot of layers/passes to make sure play with SCF correctly. I'll try to hack the IR to simulate a 4x unrolling and see how much there is to gain.

#13637 is tracking some things that may help at runtime like ensuring all allocations are queue-ordered - coupled with figuring out the issues with #13440 so we can use CUDA memory pools may be a band-aid as well. Help wanted there from a CUDA person (@jpienaar anyone available to look? mostly just debugging on a CUDA machine to figure out why the drivers are returning an error in 2 tests).

I suspect the largest thing though is going to be improvements to fusion to reduce materialization of large tensors and optimistically making dispatches in-place where possible by analysis (trivial example being elementwise consuming as the last use changing to tie the operands to the results when the sizes (not shapes) match, and linalg should help here). I have a feeling there's quite a bit of gain here but it shouldn't fall only on Mahesh - getting some extra eyes on FusionOfTensors and dispatch region formation will be critical if this is a priority - there's low hanging fruit to be had and when dealing with GB tensors even small % gains are high priority. I don't think it's a fancy cost model autotuning rabbit hole (that'd be bad) but improved heuristics and maybe a user setting (favor size vs favor speed style) - step 1 here is to get these things working and I don't think we need research projects for that.

**Comment by MaheshRavishankar** [comment]
> Damn, have some good improvements but none impact the memory consumption of Sean's particular model (with `--iree-stream-partitioning-favor=min-peak-memory` set) and instead only save a few GB of copies. The issues are still the control flow and most critically all of the out-of-place dispatches that lead to higher peak memory consumption - the transient usage there hides all of the improvements I could make at the stream level (well, not all - one transient buffer went from 1426063488 bytes to 1426063424 bytes :).
> 
> After I land these improvements (they clean up the IR quite a bit at the stream level) I can start digging into control flow to scope that out (probably "big"). #13542 mentions some things we can do but we still need to scope out the work as there's a lot of layers/passes to make sure play with SCF correctly. I'll try to hack the IR to simulate a 4x unrolling and see how much there is to gain.
> 
> #13637 is tracking some things that may help at runtime like ensuring all allocations are queue-ordered - coupled with figuring out the issues with #13440 so we can use CUDA memory pools may be a band-aid as well. Help wanted there from a CUDA person (@jpienaar anyone available to look? mostly just debugging on a CUDA machine to figure out why the drivers are returning an error in 2 tests).
> 
> I suspect the largest thing though is going to be improvements to fusion to reduce materialization of large tensors and optimistically making dispatches in-place where possible by analysis (trivial example being elementwise consuming as the last use changing to tie the operands to the results when the sizes (not shapes) match, and linalg should help here). I have a feeling there's quite a bit of gain here but it shouldn't fall only on Mahesh - getting some extra eyes on FusionOfTensors and dispatch region formation will be critical if this is a priority - there's low hanging fruit to be had and when dealing with GB tensors even small % gains are high priority. I don't think it's a fancy cost model autotuning rabbit hole (that'd be bad) but improved heuristics and maybe a user setting (favor size vs favor speed style) - step 1 here is to get these things working and I don't think we need research projects for that.

Thanks Ben! I am going to try to land somethings I have in flight, but these also are probably more cleanup than actually addressing issues... I can look into the in-place usage.  Its can get a bit tricky... Some high-level notes for future
-  we should have "in-place" behavior at `flow.dispatch.workgroups` level and not try to make the linalg ops themselves in-place.  You can still make it destination passing-style and that actually works better for the backends too. I did know there was some concern at some point about over use of "in-place", not sure why and what to balance here.
- One way I have "mimic"-ed in-place in the past is to use some sort of allocation pool (not dynamic allocation pool, but a static allocation pool)... So if a value is dead, you can free it and is usable for other values. It is not exactly in-place, but is pretty close... in-place is easier though if we can do it at `flow.dispatch.workgroups` level...

**Comment by benvanik** [comment]
The cleanups you're doing will help a lot in reducing the noise - excited for them!

In-place updates that do complete replacement are fine but partial updates can prevent concurrency (without a lot more analysis that is only probabilistically able to recover it) so partial updates are taken care of via the iree-stream-emplace-allocations pass today (well, when I send the PR that makes it 10x better :). We'll need to characterize for this model how much doing partial updates would help vs only focusing on full replacement as full replacement we can do by default while partial would be something we'd want to condition on a favor-size flag for cases where there's ambiguity. For example, if we can prove when inducing a partial in-place update that it's impossible for any other consumer to be using the value (one use on consumed value and one use on produced and such details) then we could maybe safely do them by default. Probably several steps here and we can start with the simpler cases first. slice+dispatch-in-place-into-slice+update is something we should handle today (or can make better) so maybe it'll work out in most cases 🤞  split + lots of concurrent dispatches + concat are the kind of patterns we see a lot here and we'll need to ensure survive to iree-stream-schedule-concurrency.

The allocation we do already does that but the total peak transient memory required scales with the number and size of the live values. In these models nearly all the values are live all the time (entry, during loop, exit) because of JAX returning hundreds of values instead of keeping them as state (or not returning things it doesn't care about). This means that if we have a 4GB tensor passed to the program and then returned back out a single out-of-place operation will require another 4GB. If you scale that for all the peak live tensors no amount of transient packing tetris helps - though we may be able to get a better algorithm - I'm using the bin packing one tflite has: https://github.com/google/iree/blob/main/compiler/src/iree/compiler/Dialect/Stream/Transforms/LayoutSlices.cpp/#L66-L83
Definite room for improvement there! It's still quite helpful, though, and with that algorithm we use 1426063424 / 18878605632 / 7650410496 byte transient allocations (~30GB allocated, but peak of ~18GB) and without it we use 17834258240 / 51630334848 / 28778661952 (~100GB allocated with ~51GB peak) in this particular model.

Only way to drop that is to either place things in outputs (working on it, and loop support will be important too but only goes up to the first out-of-place dispatch), improve fusions to reduce dispatch regions (fewer materialized large tensors flowing between them), and do dispatches in-place (reducing amount of live tensors at any point in time).

There's also HLO/linalg-level stuff that could also help (propagating slices up tensor chains, pushing broadcasts down to keep tensors as small as possible as long as possible, etc). I think there's several such cases that someone could find sorting through the IR and that have traditionally fallen in the "eh vector dialect/LLVM/whatever will take care of it" category but they really need to happen before dispatch region formation so that we aren't materializing the larger/more numerous tensors in the first place. Really hoping that the recent murmurs of some HLO-level optimizations pans out but having linalg ones in addition to them will be important (and always has been, but has been underinvested in - maybe now there'll be some motivation for people to contribute some things :).

For tracking, I'm currently working on:
* [ ] land improvements to iree-stream-emplace-allocations (lets us do dispatches into output tensors)
* [ ] rest of the items in #13637 (particularly making everything queue ordered allocations)
* [ ] beg someone to figure out #13440 for CUDA memory pooling of queue ordered allocations
* [ ] long-standing VM register allocation fix to ensure we drop buffers immediately
* [ ] explore removing the TopLevelSCFToCFG pass so that scf ops survive to the stream layer, then figure out what breaks there
* [ ] anything I spot after Mahesh's changes land

**Comment by allieculp** [comment]
@MaheshRavishankar can you update this issue with the latest?

**Comment by MaheshRavishankar** [comment]
This is more of a tracking bug... There is not an immediate actionable thing on this (there is a task list that I am working though). Should I change the priority somehow?

**Comment by allieculp** [comment]
I think Priority is okay, do you want to add your task list here? Just attempting to surface the next steps in some way!

**Comment by MaheshRavishankar** [comment]
Ok, Ben has a few tasks.. I have a comment above that has task list that I am working on ...
